[ONNX] STFT Support

urinieto opened the pull request:
This PR addresses issue [#81075](https://github.com/pytorch/pytorch/issues/81075),  making `torch.stft` compatible with ONNX Opset 17's STFT operator.

The conversion works for _most_ of `torch.stft` functionality:

- Batched or unbatched inputs
- Normalization
- Pre-computed windows
- Rectangular windows
- One-sided returns
- Window centering (implicitly supported)

What is currently _not_ supported is **complex types**, due to the lack of conversion functionality between PyTorch and ONNX (https://github.com/pytorch/pytorch/issues/86746).

Regardless, this is easy to bypass by setting `return_complex=False` when using `torch.stft`.

Note that there is already a draft PR to address this (https://github.com/pytorch/pytorch/pull/83944), but it is currently closed and it only partially addresses the conversion (i.e., most of `torch.stft` functionality is lacking, and unit tests are missing).

urinieto committed:
Added ONNX STFT support, including unit tests. Addressed all CR
comments.

justinchuby commented:
Thanks for your contribution! Could you sign the CLA following instructions by the bot comment above, and fix lint issues by running `lintrunner -a -m master`? I will add comments after the CLA is signed.

https://github.com/pytorch/pytorch/wiki/lintrunner

urinieto commented:
Thanks for reviewing this, @justinchuby! I just signed the CLA, and associated my last commits with the right email address. Moreover, I have passed the linter and fixed a problem with the unit tests regarding the `return_complex` parameter. Should be good to be reviewed :)

justinchuby commented on torch/onnx/symbolic_opset17.py:
```diff
@@ -54,3 +55,128 @@ def layer_norm(
+    win_length: Optional[int] = None,
+    window: Optional[_C.Value] = None,
+    normalized: bool = False,
+    onesided: Optional[bool] = True,
+    return_complex: Optional[bool] = False,
```
```suggestion
    onesided: bool = True,
    return_complex: bool = False,
```

can these values ever be None? If not:

urinieto replied:
I was following the original [`torch.stft`](https://pytorch.org/docs/stable/_modules/torch/functional.html#stft) function. While I don't think `None` makes sense here, we should make the ONNX operator as compatible with the `torch.stft` function. Thoughts?

justinchuby replied:
Sound good. thanks for pointing this out!

justinchuby commented on torch/onnx/symbolic_opset17.py:
```diff
@@ -54,3 +55,128 @@ def layer_norm(
+        Input tensor for the transformation
+    n_fft : int
+        FFT size
+    hop_length : int, optional
```
note: no need to include types as they appear in the function definition already

urinieto replied:
Done!

justinchuby commented on torch/onnx/symbolic_opset17.py:
```diff
@@ -54,3 +55,128 @@ def layer_norm(
+    # Pre-process input if needed
+    o_signal = input
+    signal_rank = symbolic_helper._get_tensor_rank(o_signal)
+    if signal_rank == 1:
```
Nice, thanks for handling all cases!

justinchuby commented on torch/onnx/symbolic_opset17.py:
```diff
@@ -54,3 +55,128 @@ def layer_norm(
+    o_window = window
+    n_win = symbolic_helper._get_tensor_dim_size(o_window, dim=0)
+    if n_win is not None:
+        assert n_win == n_fft, "Analysis window's size must the same as the FFT size"
```
Would this ever happen with valid user inputs? If not, and assertion is good but may be useful to include that this is a bug etc. If it _can_ happen with valid user inputs and it's just a case we cannot support, then raise an error.

justinchuby replied:
Potentially suggest a fix in the error message as well, if possible, everywhere where errors are raised

urinieto replied:
This only happens with non-valid user inputs. `torch.stft` also raises an error there, so I was mimicking that, not sure if it's necessary to leave it here. Happy to remove it if you think it's overkill.

I have updated all error messages with potential fixes.

justinchuby replied:
I see. If torch.stft already checks for it, I would go with an assert

urinieto replied:
Done!

justinchuby reviewed (commented):
Looks like there's still errors in the CLA. Squashing the commits may work?

urinieto commented on test/onnx/test_operators.py:
```diff
@@ -1143,6 +1143,131 @@ def test_dynamic_axes_unchange(self):
             opset_version=12,
         )
 
+    def test_stft_default(self):
```
Sounds good, thanks for the review

urinieto replied:
Just pushed the changes for this.

urinieto commented:
Apologies, I realized that the CLA should be signed under a "company contribution", since most of this work was done using my company's resources (Adobe). This contribution was internally approved before I submitted this PR, so I might need a working day or two to have someone from the Open Source Office sign the CLA. Will report asap.

milesial commented:
Not sure if related to this PR or the underlying onnxruntime, but there seems to be a off-by-one error somewhere when using graph optimization with STFT.

```
class Test(nn.Module):
    def forward(self, audio):
        stft = torch.stft(audio, 400, 160, return_complex=False)
        return stft

m = Test()
print('torch', m(torch.randn(1, 16000 * 30)).shape)

torch.onnx.export(m,
                  torch.randn(1, 16000 * 30),
                  'test.onnx',
                  export_params=True,
                  input_names=['in'],
                  output_names=['out'],
                  opset_version=17, verbose=False)

sess_options = rt.SessionOptions()
sess_options.graph_optimization_level = rt.GraphOptimizationLevel.ORT_DISABLE_ALL
sess_options.optimized_model_filepath = 'test.onnx'
session = rt.InferenceSession('test.onnx', sess_options)
print('onnx unoptimized', session.run(None, {'in': torch.randn(1, 16000 * 30).numpy()})[0].shape)

sess_options.graph_optimization_level = rt.GraphOptimizationLevel.ORT_ENABLE_BASIC
sess_options.optimized_model_filepath = 'test.onnx'
session = rt.InferenceSession('test.onnx', sess_options)
print('onnx optimized', session.run(None, {'in': torch.randn(1, 16000 * 30).numpy()})[0].shape)
```
> torch torch.Size([1, 201, 4801, 2])
> onnx unoptimized (1, 201, 4801, 2)
> onnx optimized (1, 201, 4801, 2)

> 2023-01-15 19:38:25.256968539 [W:onnxruntime:, execution_frame.cc:828 VerifyOutputSizes] Expected shape from model of {1,201,4802,2} does not match actual shape of {1,201,4801,2} for output out

The output shape matches, but the expected shape is different, and so when we add nodes on top of that it causes shape mismatch errors.

![image](https://user-images.githubusercontent.com/30204471/212563711-b6f04d2e-8700-45fe-816f-b9c517b86a79.png)

peterbell10 commented on torch/onnx/symbolic_opset17.py:
```diff
@@ -54,3 +55,128 @@ def layer_norm(
+    # Checks
+    assert (
+        return_complex is None or not return_complex
+    ), "STFT does not currently support complex types"
```
`return_complex=False` is deprecated so I don't like the idea of forcing people to use it. I see that complex64 and complex128 are mentioned as types in [this document](https://github.com/onnx/onnx/blob/main/docs/Operators.md), so is it possible to implement `view_as_real`/`view_as_complex` in onnx?

justinchuby replied:
I like the idea of having view_as_real and view_as_complex ops in onnx. Would you be willing to open an issue for a new operator? https://github.com/onnx/onnx/issues/new?assignees=&labels=operator&template=operator.md&title=

urinieto replied:
@peterbell10 I also don't like the idea of forcing `return_complex=False`, but it is an easy workaround if we don't want to wait to have complex conversion support on ONNX, which seems like it might take quite some time (see below).

@justinchuby Seems like `view_as_complex` was already requested here: https://github.com/pytorch/pytorch/issues/49793 Unfortunately, there was not enough interest for this when it was reported. I just commented on the issue, to try to bring it back.

peterbell10 replied:
Am I right in saying onnx Cast doesn't support complex types? If it did then you could do something like:
```python
def _as_real(z):
    return float(z), float(-1j * z)

def _as_complex(real, imag):
    return complex(real) + 1j * complex(imag)
```

urinieto replied:
That is correct: ONNX doesn't currently support casting for complex types (https://github.com/onnx/onnx/blob/main/docs/Operators.md#Cast).

If it did, we could simply use `Cast` on the result of `STFT` to make sure the result is returned as complex if `return_complex=True`.

That being said, would it be possible to obtain the output of `STFT`, put it on a PyTorch tensor, and convert it to complex using the `_as_complex` function above, and then put it back into the ONNX Graph?

justinchuby replied:
Yeah, otherwise there will be a graph break in onnx.

urinieto commented:
> Not sure if related to this PR or the underlying onnxruntime, but there seems to be a off-by-one error somewhere when using graph optimization with STFT.

@milesial I noticed that as well. The fact that it actually returns the correct shape in all the tests I wrote makes me think it might be a bug on the the onnxruntime API (or the ONNX STFT definition?). I couldn't find an issue reported on the `onnxruntime` repo. 

I was thinking of reporting it once this PR is merged, so that it might be easier to reproduce. Or if you wanna report it now, by all means do :)

milesial commented:
Got it, good you caught it too, I opened https://github.com/microsoft/onnxruntime/issues/14316 since I don't know if this MR is going to be merged soon

urinieto commented:
The CLA is finally signed, apologies it took so long. I'll work on the comments and feedback asap.

justinchuby commented on test/onnx/expect/TestOperators.test_stft_default.expect:
```diff
@@ -0,0 +1,115 @@
+ir_version: 8
```
I would remove all expect files as they are hard to maintain. Only

urinieto replied:
Done!

urinieto commented on torch/onnx/symbolic_opset17.py:
```diff
@@ -54,3 +55,128 @@ def layer_norm(
+    Hence, this function does not contain these two arguments.
+    See torch.stft source code for more info.
+
+    Parameters
```
Done!

urinieto commented on torch/onnx/symbolic_opset17.py:
```diff
@@ -54,3 +55,128 @@ def layer_norm(
+
+    # Remove batch dimension, if needed
+    if signal_rank == 1:
+        stft_op = g.op(
```
Done!

urinieto commented on torch/onnx/symbolic_opset17.py:
```diff
@@ -54,3 +55,128 @@ def layer_norm(
+
+    # Get STFT sizes
+    frame_step = hop_length if hop_length is not None else n_fft // 4
+    o_frame_step = g.op("Constant", value_t=torch.tensor(frame_step, dtype=torch.int64))
```
It stands for "operator", since it comes from `g.op()`. It was a way to distinguish it from `frame_step`, which is an `int`. Happy to change it!

justinchuby replied:
Perhaps `frame_step_value`, `frame_step_constant` etc.? I would just avoid any abbreviations.

urinieto replied:
Sounds good. I just updated the code to remove all variables that started with `o_`. Thanks for the suggestion!

justinchuby commented on torch/onnx/symbolic_opset17.py:
```diff
@@ -76,32 +76,24 @@ def stft(
+        window (_C.Value, optional): Analysis window. Defaults to a window of
+            all ones
+        normalized (bool, optional): Whether to return a normalized STFT
+        onesided (bool, optional): Whether to return only half (+1) of the
```
nit: no need to annotate types in the docstring because they are annotated in the function signature already

urinieto replied:
Done!

urinieto commented on torch/onnx/symbolic_opset17.py:
```diff
@@ -54,3 +55,128 @@ def layer_norm(
+        Operator for torch.stft associated with STFT (ONNX).
+    """
+    # Checks
+    assert (
```
Done!

urinieto commented on torch/onnx/symbolic_opset17.py:
```diff
@@ -54,3 +55,128 @@ def layer_norm(
+            g.op("Constant", value_t=torch.tensor([0], dtype=torch.int64)),
+        )
+    elif signal_rank > 2:
+        raise RuntimeError(
```
Done!

justinchuby commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -38,6 +38,7 @@
 from torch import Tensor
 from torch.nn.utils import rnn as rnn_utils
 from torch.onnx import _constants, verification
+from torch.onnx.errors import SymbolicValueError
```
I would import module only:

```suggestion
from torch.onnx import errors
```

and use as

 errors.SymbolicValueError

justinchuby replied:
Same in torch/onnx/symbolic_opset17.py: https://github.com/pytorch/pytorch/wiki/torch.onnx-Namespacing

urinieto replied:
Done!

justinchuby commented:
I notice the CLA bot is still having errors. Perhaps squash or rebase some commits?

justinchuby commented:
There are some build errors I believe happened because this branch is too old. Could you rebase with master?

urinieto commented:
@justinchuby , I finished addressing all the comments in the CR. I also rebased with master and squashed all commits into one. The CLA should be passing now. Please, let me know if you'd need anything else at this time. Thanks so much for all the help!

justinchuby commented:
Thank you so much! We will now let the tests to run and keep an eye on the `pull / linux-focal-py3.8-clang10-onnx` tests.

urinieto committed:
Applied linting to fix automatic tests

justinchuby commented on torch/onnx/symbolic_opset17.py:
```diff
@@ -54,3 +55,144 @@ def layer_norm(
+@_onnx_symbolic("aten::stft")
+@symbolic_helper.parse_args("v", "i", "i", "i", "v", "b", "b", "b")
+@_beartype.beartype
+def stft(
```
You may need to add "stft" to `__all__`

justinchuby commented on test/onnx/test_op_consistency.py:
```diff
@@ -302,6 +302,7 @@ def reason_flaky() -> str:
+        "stft",
         "t",
     ]
 )
```
I would add an `dont_care` in the `EXPECTED_SKIPS_OR_FAILS` list, e.g.

```
dont_care("stft", opset=opsets_before(17), ...)
```

urinieto committed:
Added stft operator to __all__. Ignoring STFT tests in ONNX Opsets below 17

justinchuby commented on test/onnx/test_op_consistency.py:
```diff
@@ -323,6 +324,7 @@ def reason_flaky() -> str:
     ),
     fixme("ceil", dtypes=[torch.float64], reason=reason_onnx_runtime_does_not_support("Ceil", ["f64"])),
     dont_care("sqrt", dtypes=BOOL_TYPES, reason=reason_onnx_does_not_support("Sqrt")),
+    dont_care("stft", opsets=opsets_before(17), reason=reason_onnx_does_not_support("STFT")),
```
```suggestion
    dont_care("stft", opsets=[opsets_before(17)], reason=reason_onnx_does_not_support("STFT")),
```

justinchuby committed:
Update test/onnx/test_op_consistency.py

justinchuby commented:
Looks like something is wrong with the skip mechanism. I’ll take a look

urinieto commented:
> Looks like something is wrong with the skip mechanism. I’ll take a look

Any updates on this? Let me know if I can help with anything else at this time :)

justinchuby commented:
Just created https://github.com/pytorch/pytorch/pull/95527

justinchuby commented:
I think we can just combine the code change and ship them here. Let me test locally

justinchuby commented on test/onnx/test_operators.py:
```diff
@@ -1143,6 +1143,131 @@ def test_dynamic_axes_unchange(self):
             opset_version=12,
         )
 
+    def test_stft_default(self):
```
Could you revert changes in this file?

urinieto committed:
Reverted test operators as requested in CR

justinchuby commented:
@urinieto I think I fixed the tests. Could you refer to https://github.com/pytorch/pytorch/pull/95531 and basically copy past the op op consistency test to this PR? I would run the test locally as well to test. It seems there are a few mismatch elements and type issues when inputs are float64:

1.

```
_ TestOnnxModelOutputConsistency_opset17CPU.test_output_match_stft_cpu_float32 (inputs='(tensor([ 8.4667,  6.0644, -3.9242, -2.2652, -8.5734, -0.1618, -6.7775, -6.9422,\n        -0.4959,  1.3513, -3.6858,  5.3404, -5.4769,  8.1663,  6.1677, -7.5895,\n        -2.2400,  0.4061,  1.3131,  2.1346,  3.5319,  0.5391, -4.3914,  4.2587,\n        -8.6332, -5.3344, -2.2530, -4.3840, -3.1485, -7.3766, -1.9144,  1.9238,\n        -5.8632, -0.4619,  6.4427, -0.9252,  0.2501, -0.7764,  1.8214,  5.7226,\n         8.5252,  5.7155,  8.5447, -0.6509, -8.0849, -4.2667,  6.1281, -0.0583,\n        -4.4734, -6.8968, -8.4227, -7.5961, -1.8255,  4.9357,  4.8658, -8.6799,\n         5.6140, -7.0426, -1.9027, -3.6493, -1.7335, -1.7671, -8.0761, -7.7709,\n        -1.4083,  0.1164, -4.0885,  3.3903, -8.1005, -0.6074,  7.9148, -3.6710,\n         8.1270,  3.2594, -8.1221,  5.6943, -1.0386, -4.0177,  7.1969, -7.2729,\n         0.9657, -1.8843,  6.4270,  2.5123,  4.3245,  3.1784, -2.1643, -1.8927,\n        -7.4167,  4.8766,  7.1458,  6.1580, -6.3484,  0.4014, -6.3444, -4.9544,\n        -5.2443,  3.0757, -5.3632, -0.1964]),)', kwargs="{'n_fft': 10, 'return_complex': False}", opset=17, sample_num=1) _
Traceback (most recent call last):
  File "/Users/justinc/Documents/GitHub/pytorch/test/onnx/test_op_consistency.py", line 440, in test_output_match
    self.run_test(model, inputs)
  File "/Users/justinc/Documents/GitHub/pytorch/test/onnx/onnx_test_common.py", line 142, in run_test
    _run_test(model, tracing_remained_onnx_input_idx)
  File "/Users/justinc/Documents/GitHub/pytorch/test/onnx/onnx_test_common.py", line 102, in _run_test
    return run_model_test(
  File "/Users/justinc/Documents/GitHub/pytorch/test/onnx/onnx_test_common.py", line 52, in run_model_test
    return verification.verify(*args, options=options, **kwargs)
  File "<@beartype(torch.onnx.verification.verify) at 0x16d3dc8b0>", line 351, in verify
  File "/Users/justinc/Documents/GitHub/pytorch/torch/onnx/verification.py", line 890, in verify
    _compare_onnx_pytorch_model(
  File "<@beartype(torch.onnx.verification._compare_onnx_pytorch_model) at 0x16d3cfa30>", line 159, in _compare_onnx_pytorch_model
  File "/Users/justinc/Documents/GitHub/pytorch/torch/onnx/verification.py", line 451, in _compare_onnx_pytorch_model
    compare_onnx_pytorch_model_with_input(input_args, input_kwargs)
  File "/Users/justinc/Documents/GitHub/pytorch/torch/onnx/verification.py", line 445, in compare_onnx_pytorch_model_with_input
    _compare_onnx_pytorch_outputs(
  File "<@beartype(torch.onnx.verification._compare_onnx_pytorch_outputs) at 0x16d3cf5b0>", line 67, in _compare_onnx_pytorch_outputs
  File "/Users/justinc/Documents/GitHub/pytorch/torch/onnx/verification.py", line 309, in _compare_onnx_pytorch_outputs
    _compare_onnx_pytorch_outputs_in_np(onnx_outs, pt_outs_np, options)
  File "<@beartype(torch.onnx.verification._compare_onnx_pytorch_outputs_in_np) at 0x16d3cf370>", line 98, in _compare_onnx_pytorch_outputs_in_np
  File "/Users/justinc/Documents/GitHub/pytorch/torch/onnx/verification.py", line 256, in _compare_onnx_pytorch_outputs_in_np
    torch.testing.assert_close(
  File "/Users/justinc/Documents/GitHub/pytorch/torch/testing/_comparison.py", line 1511, in assert_close
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 53 / 612 (8.7%)
Greatest absolute difference: 1.7615526303416118e-05 at index (5, 1, 1) (up to 1e-07 allowed)
Greatest relative difference: inf at index (1, 0, 1) (up to 0.001 allowed)
```

2.

```
_ TestOnnxModelOutputConsistency_opset17CPU.test_output_match_stft_cpu_float64 (inputs='(tensor([-5.9711,  8.5738,  3.1924,  1.1185, -0.9642,  8.4640,  0.0336,  8.4941,\n         6.6812, -4.8605,  3.8300, -0.9909,  0.6041, -6.2842,  2.8428, -8.2124,\n        -0.0116, -0.9567,  5.2923, -3.5154, -3.2262,  3.8221, -7.9327,  7.0546,\n        -8.0077,  7.0796, -3.4618,  2.2540, -8.3627, -7.5790,  3.5207, -6.8707,\n        -0.8839,  6.8797, -3.8135, -7.0665, -5.0738,  4.5949,  0.1351,  0.8092,\n        -6.7062,  4.8008,  2.7244,  2.8627, -7.4647,  1.1488, -4.2361, -7.1930,\n         4.7912,  2.4959, -7.7377,  7.5991,  5.1394,  2.9879,  1.4998,  3.9753,\n        -1.9938,  0.3298,  7.4051,  2.7719, -2.9857,  6.3974,  4.8578, -2.6819,\n         6.6250,  3.3344,  8.8279, -4.7688, -8.6279, -0.0436, -7.5067, -3.8845,\n         1.6719,  7.7874, -0.1279,  1.7595, -3.9557,  3.7888,  2.5324, -5.2899,\n        -3.7146,  5.1800, -6.0454, -5.3193, -4.1350, -1.7591,  8.5674, -6.5586,\n        -3.7816,  6.5403,  1.0214,  4.6802, -3.1810, -7.6563,  1.8006, -0.7267,\n         4.4129,  7.5974,  6.1870,  3.3346], dtype=torch.float64),)', kwargs="{'n_fft': 10, 'return_complex': False}", opset=17, sample_num=1) _
Traceback (most recent call last):
  File "/Users/justinc/Documents/GitHub/pytorch/test/onnx/test_op_consistency.py", line 440, in test_output_match
    self.run_test(model, inputs)
  File "/Users/justinc/Documents/GitHub/pytorch/test/onnx/onnx_test_common.py", line 142, in run_test
    _run_test(model, tracing_remained_onnx_input_idx)
  File "/Users/justinc/Documents/GitHub/pytorch/test/onnx/onnx_test_common.py", line 102, in _run_test
    return run_model_test(
  File "/Users/justinc/Documents/GitHub/pytorch/test/onnx/onnx_test_common.py", line 52, in run_model_test
    return verification.verify(*args, options=options, **kwargs)
  File "<@beartype(torch.onnx.verification.verify) at 0x16d3dc8b0>", line 351, in verify
  File "/Users/justinc/Documents/GitHub/pytorch/torch/onnx/verification.py", line 890, in verify
    _compare_onnx_pytorch_model(
  File "<@beartype(torch.onnx.verification._compare_onnx_pytorch_model) at 0x16d3cfa30>", line 159, in _compare_onnx_pytorch_model
  File "/Users/justinc/Documents/GitHub/pytorch/torch/onnx/verification.py", line 430, in _compare_onnx_pytorch_model
    onnx_session = _onnx_backend_session(onnx_model_f, options.backend)
  File "<@beartype(torch.onnx.verification._onnx_backend_session) at 0x16d3cf250>", line 52, in _onnx_backend_session
  File "/Users/justinc/Documents/GitHub/pytorch/torch/onnx/verification.py", line 227, in _onnx_backend_session
    onnx_session = _ort_session(model, (backend.value,))
  File "<@beartype(torch.onnx.verification._ort_session) at 0x16d3cef80>", line 63, in _ort_session
  File "/Users/justinc/Documents/GitHub/pytorch/torch/onnx/verification.py", line 197, in _ort_session
    ort_session = onnxruntime.InferenceSession(
  File "/Users/justinc/Documents/GitHub/pytorch/venv/lib/python3.10/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 347, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "/Users/justinc/Documents/GitHub/pytorch/venv/lib/python3.10/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 386, in _create_inference_session
    sess = C.InferenceSession(session_options, self._model_bytes, False, self._read_config_from_model)
onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Type Error: Type parameter (T1) of Optype (STFT) bound to different types (tensor(double) and tensor(float) in node (/STFT).
```

justinchuby commented:
You may run the test with `pytest test/onnx/test_op_consistency.py -v -s -k stft `

urinieto committed:
Addressing problem with missmatch types between input and window

urinieto committed:
Copying changes in tests done in #95531

urinieto commented:
> You may run the test with `pytest test/onnx/test_op_consistency.py -v -s -k stft`

Hey @justinchuby , I fixed the problem regarding the type mismatch [here](https://github.com/pytorch/pytorch/pull/92087/commits/a8724c39a57bb18bc38486d09dea4c48e9989441). 

The other problem (i.e., the small differences between the results), should be resolved setting the tolerance to around 1e-5 (like I did, e.g., [here](https://github.com/pytorch/pytorch/blob/baa71292205c89a9d14708531a2c5e8dc880ac7a/test/onnx/test_pytorch_onnx_onnxruntime.py#L7583)). I'm not familiar with how to change this tolerance in the `test_op_consistency.py` file, could you take care of it?

For reference, here's the last chunk of my output of `pytest test/onnx/test_op_consistency.py -v -s -k stft`:

```bash
FAILED test/onnx/test_op_consistency.py::TestOnnxModelOutputConsistency_opset17CPU::test_output_match_stft_cpu_float32 - AssertionError: Tensor-likes are not close!
FAILED test/onnx/test_op_consistency.py::TestOnnxModelOutputConsistency_opset17CPU::test_output_match_stft_cpu_float64 - AssertionError: Tensor-likes are not close!
FAILED test/onnx/test_op_consistency.py::TestOnnxModelOutputConsistency_opset18CPU::test_output_match_stft_cpu_float32 - AssertionError: Tensor-likes are not close!
FAILED test/onnx/test_op_consistency.py::TestOnnxModelOutputConsistency_opset18CPU::test_output_match_stft_cpu_float64 - AssertionError: Tensor-likes are not close!
```

Thanks for your help!

justinchuby commented:
Thanks! I will work on it next week when I am back

justinchuby committed:
Adjust tolerance and format

justinchuby committed:
Format

justinchuby reviewed (approved):
Thanks so much for creating this!

justinchuby commented:
@pytorchbot merge -g

justinchuby committed:
Tolerance

justinchuby commented:
@pytorchbot merge -g

justinchuby committed:
Update test_op_consistency.py

justinchuby committed:
Update symbolic_opset17.py

justinchuby commented:
@pytorchbot merge

justinchuby commented:
@pytorchbot merge -f "unrelated cuda and other failures"

urinieto commented:
I am trying to export a PyTorch model with STFT into ONNX, but seems like the main torch branch does not support this (any more? yet?). Can anyone explain what happened to this PR? I basically had to repatch this PR to make the export process again. /cc @justinchuby

justinchuby commented:
It should be merged properly? Was the code gone?

urinieto commented:
This PR should be able to export to ONNX models with STFT-related methods like the following:

```python
import torch
import torchaudio
 
ONNX_MODEL = "out_melspec.onnx"
N_SAMPLES = 16000
N_FFT = 1024
OPSET_VERSION = 17
 
class CustomMelSpec(torch.nn.Module):
    def forward(self, x):
        return  torchaudio.transforms.MelSpectrogram(n_fft=N_FFT)(x)
 
x = torch.randn(1, 1, N_SAMPLES)
with open(ONNX_MODEL, "wb") as f:
    torch.onnx.export(CustomMelSpec(), (x), f, opset_version=OPSET_VERSION)
```

But running this script returns the following error (see pytorch and torchaudio versions used below):

```
PyTorch Version: 2.3.1+cu121
Torchaudio Version: 2.3.1+cu121
/opt/conda/envs/audiotagger2/lib/python3.9/site-packages/torch/onnx/_internal/jit_utils.py:307: UserWarning: Constant folding - Only steps=1 can be constant folded fo
r opset >= 10 onnx::Slice op. Constant folding not applied. (Triggered internally at ../torch/csrc/jit/passes/onnx/constant_fold.cpp:179.)
  _C._jit_pass_onnx_node_shape_type_inference(node, params_dict, opset_version)
Traceback (most recent call last):
  File "/sensei-fs/users/onieto/audiotagger2/caca.py", line 20, in <module>
    torch.onnx.export(CustomMelSpec(), (x), f, opset_version=OPSET_VERSION)
  File "/opt/conda/envs/audiotagger2/lib/python3.9/site-packages/torch/onnx/utils.py", line 516, in export
    _export(
  File "/opt/conda/envs/audiotagger2/lib/python3.9/site-packages/torch/onnx/utils.py", line 1612, in _export
    graph, params_dict, torch_out = _model_to_graph(
  File "/opt/conda/envs/audiotagger2/lib/python3.9/site-packages/torch/onnx/utils.py", line 1138, in _model_to_graph
    graph = _optimize_graph(
  File "/opt/conda/envs/audiotagger2/lib/python3.9/site-packages/torch/onnx/utils.py", line 677, in _optimize_graph
    graph = _C._jit_pass_onnx(graph, operator_export_type)
  File "/opt/conda/envs/audiotagger2/lib/python3.9/site-packages/torch/onnx/utils.py", line 1956, in _run_symbolic_function
    return symbolic_fn(graph_context, *inputs, **attrs)
  File "/opt/conda/envs/audiotagger2/lib/python3.9/site-packages/torch/onnx/symbolic_helper.py", line 306, in wrapper
    return fn(g, *args, **kwargs)
  File "/opt/conda/envs/audiotagger2/lib/python3.9/site-packages/torch/onnx/symbolic_opset17.py", line 115, in stft
    raise errors.SymbolicValueError(
torch.onnx.errors.SymbolicValueError: STFT does not currently support complex types  [Caused by the value '73 defined in (%73 : Float(*, *, strides=[17024, 1], requires_grad=0, device=cpu) = onnx::Reshape[allowzero=0](%63, %72), scope: __main__.CustomMelSpec:: # /opt/conda/envs/audiotagger2/lib/python3.9/site-packages/torch/functional.py:664:0
```

After applying this PR as a patch again (to torch v1.13), the code runs fine and I can successfully export the model. So I've no idea what happened to the merged code 🤔 Maybe the way to export ONNX models has changed?

urinieto commented:
Any updates on this, @justinchuby ?

urinieto commented:
Yeah, I have no idea why this is not working right off the bat with the latest torch version (or basically any version after this patch was merged). For whatever reason, the STFT models are not converted to ONNX any more :/

justinchuby commented:
The error message is saying the inputs cannot be complex; we can only run them in the real value mode. Was this what the model was doing?

urinieto commented:
After some more digging, turns out the torch code is totally fine, it was a problem with torchaudio!

Everything works fine after applying this patch (basically forcing torchaudio spectrogram to use float values instead of complex types):

```patch
diff --git a/src/torchaudio/functional/functional.py b/src/torchaudio/functional/functional.py
index af34e707..4eb842c7 100644
--- a/src/torchaudio/functional/functional.py
+++ b/src/torchaudio/functional/functional.py
@@ -133,9 +134,12 @@ def spectrogram(
         pad_mode=pad_mode,
         normalized=frame_length_norm,
         onesided=onesided,
-        return_complex=True,
+        return_complex=False,
     )

+    # From imaginary and real values to absolute value
+    spec_f = torch.sqrt(torch.pow(spec_f[:, :, :, 0], 2.0) + torch.pow(spec_f[:, :, :, 1], 2.0))
+
     # unpack batch
     spec_f = spec_f.reshape(shape[:-1] + spec_f.shape[-2:])
```

Apologies for the confusion, and thanks for helping me figure this out!
